controllers/jenkins/pipelinerun/pipelinerun_controller.go

```diff
@@ -25,14 +25,33 @@ import (
 	"github.com/jenkins-zh/jenkins-client/pkg/job"
 	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/tools/record"
 	"k8s.io/client-go/util/retry"
 	"kubesphere.io/devops/pkg/api/devops/v1alpha3"
 	devopsv1alpha4 "kubesphere.io/devops/pkg/api/devops/v1alpha4"
 	devopsClient "kubesphere.io/devops/pkg/client/devops"
 	"reflect"
 	ctrl "sigs.k8s.io/controller-runtime"
 	"sigs.k8s.io/controller-runtime/pkg/client"
-	time "time"
+	"time"
+)
+
+// EventType is type of event.
+type EventType string
+
+// EventReason is reason of event.
+type EventReason string
+
+const (
+	Started        EventReason = "Stareted"
+	Updated        EventReason = "Updated"
+	TriggerFailed  EventReason = "TriggerFailed"
+	RetrieveFailed EventReason = "RetrieveFailed"
```

Comment:
You don't need to convert `EventReason` to `string` if you don't create a type alias. Such as:

```
const (
  Updated string = "Updated"
)
```

See also:

* https://github.com/kubernetes/kubernetes/blob/0b4a793da2a2912393687367e0af2436612a9b8e/staging/src/k8s.io/api/core/v1/types.go#L5488
* https://github.com/kubesphere/ks-devops/blob/db195546d263e5fafa6829108bbca8ea982adc34/controllers/devopsproject/devopsproject_controller.go#L284

Reply:
That's great.